bitbybit/src/bitenum.rs

```diff
@@ -1,303 +1,282 @@
-use proc_macro::TokenStream;
-use std::str::FromStr;
+use std::fmt;
 
-use proc_macro2::TokenTree;
-use quote::{quote, ToTokens};
-use syn::__private::TokenStream2;
-use syn::{Attribute, Data, DeriveInput, Expr, Ident};
+use proc_macro2::{Span, TokenStream};
+use quote::{quote, spanned::Spanned, ToTokens};
+use syn::{meta::ParseNestedMeta, Token};
 
-pub fn bitenum(args: TokenStream, input: TokenStream) -> TokenStream {
-    let args: Vec<_> = proc_macro2::TokenStream::from(args).into_iter().collect();
+use crate::bit_size::Bits;
 
-    let mut bits: Option<usize> = None;
-    let mut exhaustive_value: Option<TokenStream2> = None;
-
-    enum ArgumentType {
-        Exhaustive,
-    }
-    let mut next_expected: Option<ArgumentType> = None;
+enum Error<'a> {
+    MissingSize,
+    InvalidExhaustive,
+    InvalidAttribute,
+    InvalidExhaustiveNextToken,
+    NotConditional,
+    NotExhaustive {
+        count: u32,
+        size: usize,
+    },
+    Exhaustive {
+        count: u32,
+        max_count: u128,
+    },
+    TooManyVariants {
+        count: u32,
+        max_count: u128,
+    },
+    MissingDiscriminant {
+        variant: &'a syn::Ident,
+        suggest: u128,
+    },
+    NonLitDiscriminant {
+        variant: &'a syn::Ident,
+        suggest: u128,
+    },
+    TooLargeDiscriminant {
+        max_value: u128,
+    },
+}
 
-    fn handle_next_expected(
-        next_expected: &Option<ArgumentType>,
-        default_value: &mut Option<TokenStream2>,
-        token_stream: TokenStream2,
-    ) {
-        match next_expected {
-            None => panic!("bitenum!: Seen {}, but didn't expect anything. Example of valid syntax: #[bitenum(u3, exhaustive = false)]", token_stream),
-            Some(ArgumentType::Exhaustive) => {
-                *default_value = Some(token_stream);
-            }
+impl fmt::Display for Error<'_> {
```

Comment:
This could be replaced by a `thiserror::Error` derive macro. But it does add a fairly slow-to-compile dependency, which isn't ideal when used in a proc macro.